core/exchange_controller.go

```diff
@@ -3,61 +3,118 @@ package core
 import (
 	"errors"
 	"math/big"
+	"sort"
 
 	"github.com/dominant-strategies/go-quai/common"
-	"github.com/dominant-strategies/go-quai/common/logistic"
 	"github.com/dominant-strategies/go-quai/consensus/misc"
 	"github.com/dominant-strategies/go-quai/core/rawdb"
 	"github.com/dominant-strategies/go-quai/core/types"
+	"github.com/dominant-strategies/go-quai/log"
 	"github.com/dominant-strategies/go-quai/params"
 )
 
-// CalculateExchangeRate takes in the parent block and the etxs generated in the
-// current block and calculates the new exchange rate
-func CalculateExchangeRate(hc *HeaderChain, block *types.WorkObject, newTokenChoiceSet types.TokenChoiceSet) (*big.Int, *big.Float, *big.Float, error) {
+func CalculateBetaFromMiningChoiceAndConversions(hc *HeaderChain, block *types.WorkObject, parentExchangeRate *big.Int, newTokenChoiceSet types.TokenChoiceSet) (*big.Int, error) {
 
-	// Read the parents beta values
-	betas := rawdb.ReadBetas(hc.headerDb, block.Hash())
-	if betas == nil {
-		return nil, nil, nil, errors.New("could not find the betas stored for parent hash")
+	// Until there are tokenChoicesSetSize of miner token choices, the exchange rate is unchanged
+	if block.NumberU64(common.PRIME_CTX) < params.ControllerKickInBlock+types.C_tokenChoiceSetSize {
+		return params.ExchangeRate, nil
 	}
 
-	if block.NumberU64(common.ZONE_CTX) < types.C_tokenChoiceSetSize {
-		return params.ExchangeRate, betas.Beta0(), betas.Beta1(), nil
+	var totalQiChoices uint64 = 0
+	var totalQuaiChoices uint64 = 0
+
+	tokenChoicesSet := make([]types.TokenChoices, 0)
+	for _, tokenChoices := range newTokenChoiceSet {
+		totalQuaiChoices += tokenChoices.Quai
+		totalQiChoices += tokenChoices.Qi
+		tokenChoicesSet = append(tokenChoicesSet, tokenChoices)
+
+		if tokenChoices.Diff.Cmp(common.Big0) == 0 {
+			panic("token choices diff is zero for the first time")
```

Comment:
Ideally we never panic in any consensus code

Reply:
i can remove all these panics